### PR TITLE
openvscode-server: add NixOS module and tests

### PR DESCRIPTION
**Pull request by drupol**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        description = lib.mdDoc ''
+          Which openvscode-server derivation to use.
+        '';
+        type = types.package;
+      };
+
+      extraPackages = mkOption {
+        default = [ ];
+        description = lib.mdDoc ''
+          Packages that are available in the PATH of openvscode-server.
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:25]:
```suggestion
          Additional packages to add to the openvscode-server {env}`PATH`.
```

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        '';
+        example = "[ pkgs.go ]";
+        type = types.listOf types.package;
+      };
+
+      extraEnvironment = mkOption {
+        type = types.attrsOf types.str;
+        description = lib.mdDoc ''
+          Additional environment variables to passed to openvscode-server.
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:34]:
```suggestion
          Additional environment variables to pass to openvscode-server.
```

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        '';
+        default = { };
+        example = { PKG_CONFIG_PATH = "/run/current-system/sw/lib/pkgconfig"; };
+      };
+
+      extraArguments = mkOption {
+        default = [ ];
+        description = lib.mdDoc ''
+          Additional arguments that passed to openvscode-server
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:43]:
```suggestion
          Additional arguments to pass to openvscode-server.
```

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        '';
+        example = ''[ "--log=info" ]'';
+        type = types.listOf types.str;
+      };
+
+      host = mkOption {
+        default = "127.0.0.1";
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:50]:
```suggestion
        default = "localhost";
```

I think we should use the same default host as upstream.

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        description = lib.mdDoc ''
+          The host name or IP address the server should listen to.
+        '';
+        type = types.str;
+      };
+
+      port = mkOption {
+        default = 4444;
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:58]:
```suggestion
        default = 3000;
```

I think we should use the same default port as upstream.

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      enable = mkEnableOption (lib.mdDoc "openvscode-server");
+
+      package = mkOption {
+        default = pkgs.openvscode-server;
+        defaultText = lib.literalExpression "pkgs.openvscode-server";
+        description = lib.mdDoc ''
+          Which openvscode-server derivation to use.
+        '';
+        type = types.package;
+      };
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:20]:
```suggestion
      package = mkPackageOptionMD pkgs "openvscode-server" { };
```

**Review by felschr — CHANGES_REQUESTED** [review]
Module needs to be referenced in `nixos/modules/module-list.nix`.
Can you also add a test?

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+            --user-data-dir=${cfg.user-data-dir} \
+          '' + optionalString (cfg.server-data-dir != null) ''
+            --server-data-dir=${cfg.server-data-dir} \
+          '' + optionalString (cfg.extensions-dir != null) ''
+            --extensions-dir=${cfg.extensions-dir} \
+          '' + optionalString (cfg.connection-token != null) ''
+            --connection-token=${cfg.connection-token} \
+          '' + optionalString (cfg.connection-token-file != null) ''
+            --connection-token-file=${cfg.connection-token-file} \
+          '' + lib.escapeShellArgs cfg.extraArguments;
```
Comment (felschr) [inline nixos/modules/services/web-apps/openvscode-server.nix:186]: Do we need all these options?
I guess it makes sense to have options for arguments that are almost always used, but for the rest `extraArguments` should be enough. The example for `extraArguments` could also include some common options.

Comment (drupol) [reply nixos/modules/services/web-apps/openvscode-server.nix:186]: I agree, but I think those are the most important. I don't think we should change that.

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        example = [ "docker" ];
+        type = types.listOf types.str;
+      };
+
+      withoutConnectionToken = mkOption {
+        default = false;
+        description = lib.mdDoc ''
+          Run without a connection token. Only use this if the connection is secured by other means.
+        '';
+        example = "true";
```
Comment (IndeedNotJames) [inline nixos/modules/services/web-apps/openvscode-server.nix:92]:
```suggestion
        example = true;
```

**Review by emilylange — CHANGES_REQUESTED** [review]
Please also add your nixos VM test to the package's `passthru.tests` like so:

```diff
diff --git a/pkgs/servers/openvscode-server/default.nix b/pkgs/servers/openvscode-server/default.nix
index 531e08e7cd0..f5c17fcaff6 100644
--- a/pkgs/servers/openvscode-server/default.nix
+++ b/pkgs/servers/openvscode-server/default.nix
@@ -1,7 +1,7 @@
 { lib, stdenv, fetchFromGitHub, buildGoModule, makeWrapper
 , cacert, moreutils, jq, git, pkg-config, yarn, python3
 , esbuild, nodejs_16, libsecret, xorg, ripgrep
-, AppKit, Cocoa, Security, cctools }:
+, AppKit, Cocoa, Security, cctools, nixosTests }:
 
 let
   system = stdenv.hostPlatform.system;
@@ -164,6 +164,10 @@ in stdenv.mkDerivation rec {
     ln -s ${nodejs}/bin/node $out
   '';
 
+  passthru.tests = {
+    inherit (nixosTests) openvscode-server;
+  };
+
   meta = with lib; {
     description = "Run VS Code on a remote machine";
     longDescription = ''
```

---

And please consider rephrasing your commits messages to

`nixos/openvscode-server: init` and
`nixosTests.openvscode-server: init` (or `nixos/tests/openvscode-server: init`)

---

And add an entry to https://github.com/NixOS/nixpkgs/blob/5ff59e34f93324297e18009bb27e81e8cd46f0d2/nixos/doc/manual/release-notes/rl-2305.section.md

---

There is also #229030, but I don't think there is an RFC enforcing it yet.

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+      enable = mkEnableOption (lib.mdDoc "openvscode-server");
+
+      package = mkPackageOptionMD pkgs "openvscode-server" { };
+
+      extraPackages = mkOption {
+        default = [ ];
+        description = lib.mdDoc ''
+          Additional packages to add to the openvscode-server {env}`PATH`.
+        '';
+        example = "[ pkgs.go ]";
```
[the remark this answers is not in the document]
Comment (drupol) [reply nixos/modules/services/web-apps/openvscode-server.nix]: Fixed.

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        default = { };
+        example = { PKG_CONFIG_PATH = "/run/current-system/sw/lib/pkgconfig"; };
+      };
+
+      extraArguments = mkOption {
+        default = [ ];
+        description = lib.mdDoc ''
+          Additional arguments to pass to openvscode-server.
+        '';
+        example = ''[ "--log=info" ]'';
```
[the remark this answers is not in the document]
Comment (drupol) [reply nixos/modules/services/web-apps/openvscode-server.nix]: Fixed.

File: nixos/modules/services/web-apps/openvscode-server.nix (hunk 8 above)
Comment (drupol) [reply nixos/modules/services/web-apps/openvscode-server.nix]: Fixed.

File: pkgs/servers/openvscode-server/default.nix [hunk 11]
```diff
@@ -164,15 +164,19 @@ in stdenv.mkDerivation rec {
     ln -s ${nodejs}/bin/node $out
   '';
 
-  meta = with lib; {
+  passthru.tests = {
+    inherit (nixosTests) openvscode-server;
+  };
+
+  meta = {
```
Comment (IndeedNotJames) [inline pkgs/servers/openvscode-server/default.nix:171]: `with lib;` in the package's `meta` is fine.
The mentioned #229030 seems to care only about `with lib;` in nixos modules, not packages.

Please undo :)
And sorry for the confusion^^

Comment (drupol) [reply pkgs/servers/openvscode-server/default.nix]: Fixed!

File: pkgs/servers/openvscode-server/default.nix [hunk 12]
```diff
@@ -164,6 +164,10 @@ in stdenv.mkDerivation rec {
     ln -s ${nodejs}/bin/node $out
   '';
 
+  passthru.tests = {
+    inherit (nixosTests) openvscode-server;
+  };
```
Comment (IndeedNotJames) [inline pkgs/servers/openvscode-server/default.nix:169]: This file is missing `nixosTests` in lines 1-4.
Please add it :)

```patch
diff --git a/pkgs/servers/openvscode-server/default.nix b/pkgs/servers/openvscode-server/default.nix
index f0133633fcd..f5c17fcaff6 100644
--- a/pkgs/servers/openvscode-server/default.nix
+++ b/pkgs/servers/openvscode-server/default.nix
@@ -1,7 +1,7 @@
 { lib, stdenv, fetchFromGitHub, buildGoModule, makeWrapper
 , cacert, moreutils, jq, git, pkg-config, yarn, python3
 , esbuild, nodejs_16, libsecret, xorg, ripgrep
-, AppKit, Cocoa, Security, cctools }:
+, AppKit, Cocoa, Security, cctools, nixosTests }:
 
 let
   system = stdenv.hostPlatform.system;
```

**Review by emilylange — CHANGES_REQUESTED** [review]

File: pkgs/servers/openvscode-server/default.nix (hunk 12 above)
Comment (drupol) [reply pkgs/servers/openvscode-server/default.nix:169]: Argh ! Just did.

**Comment by IndeedNotJames** [comment]
@ofborg test openvscode-server

**Review by emilylange — APPROVED** [review]

**Review by felschr — APPROVED** [review]

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+  config = lib.mkIf cfg.enable {
+    systemd.services.openvscode-server = {
+      description = "OpenVSCode server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      path = cfg.extraPackages;
+      environment = cfg.extraEnvironment;
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/openvscode-server \
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/openvscode-server.nix:165]: Maybe use lib.getExe

File: nixos/modules/services/web-apps/openvscode-server.nix [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        default = false;
+        description = lib.mdDoc ''
+          Run without a connection token. Only use this if the connection is secured by other means.
+        '';
+        example = true;
+        type = lib.types.bool;
+      };
+
+      socketPath = lib.mkOption {
+        default = null;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/openvscode-server.nix:96]:
```suggestion
        default = null;
        example = "/run/openvscode/socket";
```

File: nixos/modules/services/web-apps/openvscode-server.nix (hunk 13 above)
Comment (drupol) [reply nixos/modules/services/web-apps/openvscode-server.nix]: Fixed.

**Pull request merged**